### PR TITLE
[v7r1] DFC FileManagerPS: getReplicas speed improvment

### PR DESCRIPTION
**Pull request by chaen**
Not legally binding measurements, but still:
Before
`Getting replicas for 45540 LFNs : completed in 130.1 seconds`
After
`Getting replicas for 45540 LFNs : completed in 15.7 seconds`

It requires an update to the DB, so I'll add a note to the wiki

BEGINRELEASENOTES
*DMS
CHANGE: FileManagerPS has bulk method for getReplicas

ENDRELEASENOTES

### CODE REVIEW

File: DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py [hunk 1]
```diff
@@ -775,51 +775,53 @@ def _setFileParameter(self, fileID, paramName, paramValue, connection=False):
...
+    # always add Status in the list of required fields
     if 'Status' not in fields:
       fields.append('Status')
 
-    replicas = {}
+    # We initialize the dictionnary with empty dict
+    # as default value, because this is what we want for
+    # non existing replicas
+    # note (do NOT use dict.fromkeys, otherwise you get
+    # the same dict for all the fileIDs)
```
Comment (chrisburr) [inline DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:801]: Do we need comments like this? 
```suggestion
```
Also I think this belongs just before `for row in rows:` to make clear that is isn't actually used until then.

Comment (chaen) [reply DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:801]: What should go before `for row in rows` ? The `replicas` declaration or the comment ?

Comment (chrisburr) [reply DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:801]: The replicas declaration

Comment (chaen) [reply DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:801]: it's not used before the `for row` part, but I don't want it to be reinitialized for every chunk.
I can also remove the comment, but if anyone ever changes that to a `fromdict` (including me :-p ), you owe me a huge beer :-D

Comment (chrisburr) [reply DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py]: Ah sorry I missed the chunked for loop 🤦

**Review by chrisburr — DISMISSED** [review]

File: DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py [hunk 2]
```diff
@@ -775,51 +775,51 @@ def _setFileParameter(self, fileID, paramName, paramValue, connection=False):
...
+      fields_input = ['PFN']
 
     fields = list(fields_input)
 
+    # always add Status in the list of required fields
     if 'Status' not in fields:
       fields.append('Status')
 
-    replicas = {}
+    # We initialize the dictionnary with empty dict
```
Comment (fstagni) [inline DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:797]:
```suggestion
    # We initialize the dictionary with empty dict
```
:smirk:

File: DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py [hunk 3]
```diff
@@ -775,51 +775,51 @@ def _setFileParameter(self, fileID, paramName, paramValue, connection=False):
...
+    # as default value, because this is what we want for
+    # non existing replicas
+    replicas = {fileID: {} for fileID in fileIDs}
```
Comment (fstagni) [inline DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:800]: what if you use `collections.defaultdict(dict)` ?

File: DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py [hunk 4]
```diff
@@ -775,51 +775,51 @@ def _setFileParameter(self, fileID, paramName, paramValue, connection=False):
...
     fieldNames = ["FileID", "SE", "Status", "RepType", "CreationDate", "ModificationDate", "PFN"]
 
-    for fileID in fileIDs:
-
-      result = self.db.executeStoredProcedureWithCursor('ps_get_all_info_of_replicas',
-                                                        (fileID, allStatus, fStatus))
+    for chunks in breakListIntoChunks(fileIDs, 1000):
+      # Format the FileIDs to be used in a IN clause in the stored procedure
+      formatedFileIds = intListToString(chunks)
+      result = self.db.executeStoredProcedureWithCursor('ps_get_all_info_of_replicas_bulk',
```
Comment (fstagni) [inline DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:810]: Changes to the DBs should be done only for new releases. Here you end up imposing the deploy of a new stored procedure in patch 40 of v7r1. It's not correct.

My suggestion is that here you try the `bulk` procedure and fall back to the old one if not present. Then we repeat in the v7r2 notes about deploying the new `bulk` procedure.

File: DataManagementSystem/DB/FileCatalogWithFkAndPsDB.sql [hunk 5]
```diff
@@ -1364,6 +1365,52 @@ BEGIN
...
+
+  ELSE
+
+    SET @sql = CONCAT(
+                    'SELECT SQL_NO_CACHE FileID, se.SEName, st.Status, RepType, CreationDate, ModificationDate, PFN
+                    FROM FC_Replicas r
+                    JOIN FC_StorageElements se on r.SEID = se.SEID
+                    JOIN FC_Statuses st on r.Status = st.StatusID
+                    WHERE FileID IN (', file_ids, ') ',
+                    'and st.Status  in (',visibleReplicaStatus,') ');
```
Comment (fstagni) [inline DataManagementSystem/DB/FileCatalogWithFkAndPsDB.sql:1402]: not that it makes a difference, but for being consistent with the formatting:

```suggestion
                    'AND st.Status IN (', visibleReplicaStatus, ') ');
```

File: DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py (hunk 3 above)
Comment (chrisburr) [reply DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:800]: Then files without replicas won't be included in the result.

File: DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py (hunk 2 above)
Comment (chaen) [reply DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:797]: ... shall we put "correctly spelling dictionnary" as an objective for my next MERIT ? :laughing:

File: DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py (hunk 3 above)
Comment (chaen) [reply DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:800]: yes, exactly

File: DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py (hunk 4 above)
Comment (chaen) [reply DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:810]: In general I totally agree. However this is for the LHCb DFC that I think absolutely nobody but use uses... I'll have a quick look if I can do that in a neat way but I wouldn't bother too much... I may as well move all these managers in LHCbDIRAC

File: DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py (hunk 2 above)
Comment (fstagni) [reply DataManagementSystem/DB/FileCatalogComponents/FileManager/FileManagerPs.py:797]: no, we should always put attainable objectives 😄

File: DataManagementSystem/DB/FileCatalogWithFkAndPsDB.sql (hunk 5 above)
Comment (chaen) [reply DataManagementSystem/DB/FileCatalogWithFkAndPsDB.sql:1402]: yep, done thanks

**Comment by atsareg** [comment]
I think this is OK for the patch only because it is solely used by LHCb.

**Review by atsareg — APPROVED** [review]

**Pull request merged**